Customizable nix installer

FriedrichAltheide opened the pull request:
calamares-nixos-extensions does not currently allow you to easily create a custom installer with a customized configuration.nix.
This PR adds the idea of snippets (#15), where many of the previously defined variables containing config parts are moved into individual snippet files.
All default snippets are in the `defaultConfigs/snippets` directory. If you want to create a custom installer, simply place a copy of the snippet - with your changes - in the new `customConfigs/snippets' folder.
It will then be used instead of the ones defined in this repository.

As well as snippets, modules can be defined by copying them into the `customConfigs/modules` folder. Note that if the module is also to be imported, the imports.snippet must be customised using the method above.

**Try this PR together with https://github.com/FriedrichAltheide/nixpkgs/commits/calamares-nixos-extensions-custom-installer/**

FriedrichAltheide commented:
@vlinkz Would you be so kind, and give me a hint, if it is worth the time to rebase this PR-draft/if the idea of the PR is even wanted?

FriedrichAltheide commented:
@vlinkz ping

vlinkz commented:
Yeah definitely seems like a good idea. I'll try to take a deeper look tomorrow or later this week

FriedrichAltheide commented:
The code has just been rebased and has not yet been tested again

FriedrichAltheide committed:
Allow building installers with custom configs

    - Move previous config snippets into snippet files, used as default snippets (getCustomOrDefaultSnippet)
    - If a custom snippet is found in customConfigSnippets, it is used instead

FriedrichAltheide commented:
An initial draft of a corresponding NixOS module can be found here: https://github.com/FriedrichAltheide/nixpkgs/commits/calamares-nixos-extensions-custom-installer/

FriedrichAltheide committed:
Adding a customized calamares installer example

FriedrichAltheide commented:
Is there any initial feedback yet?

FriedrichAltheide commented:
@vlinkz ping

muellerbernd commented on modules/nixos/main.py:
```diff
@@ -672,13 +397,13 @@ def run():
         and gs.value("packagechooser_packagechooser") != ""
     ):
-        cfg += cfgmisc
+        cfg += getCustomOrDefaultSnippet('misc')
```
in contrast to the original `cfgmisc` the audio settings are missing in `misc` snippet.
```suggestion
        cfg += getCustomOrDefaultSnippet('misc')
        cfg += getCustonOrDefaultSnippet('audio')
```

muellerbernd reviewed (approved):
I like these changes and they make the code much cleaner. Tested the provided flake example for a custom calamares ISO with additional changes for my needs and it worked like expected. Would be nice if this PR gets merged.

FriedrichAltheide committed:
Load audio snippet

@muellerbernd thanks for spotting

Co-authored-by: Bernd Müller <35687881+muellerbernd@users.noreply.github.com>

daniel-xoul-ai commented:
Just played around with this PR. Sorry if this is the wrong forum / inappropriate (newbie to NixOS), but is there a way to override/disable certain snippets? In my case, I want to disable showing the "user" creation step in Calamares, but it's not clear if this PR provides a way to do it.

Also huge fan of the example-custom-iso, really helped get me up to speed.

FriedrichAltheide commented:
@daniel-xoul-ai have a look at the unfree snippet in example-custom-iso/custom-calamares/default.nix
There you can see that you can overwrite an snippet with an empty one: `unfree = '''';`

FriedrichAltheide commented on example-custom-iso/custom-calamares/default.nix:
```diff
@@ -0,0 +1,82 @@
+        ../modules/virtualboxGuest.nix
+      ];
+
+      unfree = '''';
```
```suggestion
      # default snippets can also be removed, by setting the snippet to an empty string
      unfree = '''';
```

FriedrichAltheide committed:
add comment how to remove a default snippet

FriedrichAltheide commented:
@vlinkz any updates?

FriedrichAltheide committed:
Fix function name

vlinkz commented on modules/nixos/main.py:
```diff
@@ -805,6 +533,12 @@ def run():
     # Write the configuration.nix file
     libcalamares.utils.host_env_process_output(["cp", "/dev/stdin", config], None, cfg)
+    # Format configuration file
+    libcalamares.utils.host_env_process_output(["nixfmt",  config], None)
```
Right now `nixfmt` is not in the build ISO so installation fails

FriedrichAltheide replied:
Have you applied my proposed module changes before building the iso? https://github.com/FriedrichAltheide/nixpkgs/commits/calamares-nixos-extensions-custom-installer/

emilazy replied:
Is there a reason we can’t just ensure the templates we ship are formatted so that the resulting config will be without an explicit formatting step?

FriedrichAltheide replied:
It would be possible to format the default snippets in a way that the default configuration would look nicely. Yet, this PR intends to add an easy way to build custom installers based on the default calamares installer (e.g, some default settings like proxy, SSO, or CA settings). Hence, it would be nice to format the generated configuration.nix, ensuring that it is formatted nicely for new users.

vlinkz commented on modules/nixos/main.py:
```diff
@@ -805,6 +533,12 @@ def run():
+    libcalamares.utils.host_env_process_output(["nixfmt",  config], None)
+
+    # Copy modules
+    copyModules(root_mount_point, 'defaultConfigs/modules')
```
Right now this leaves an empty `modules` folder in a users `/etc/nixos` dir. Maybe we should remove [modules/nixos/defaultConfigs/modules](https://github.com/NixOS/calamares-nixos-extensions/tree/39f4fdc00c111bf506c43d1025b914dbd43192da/modules/nixos/defaultConfigs/modules) directory and add a check to `copyModules` to return if the directory doesn't exist?

FriedrichAltheide replied:
I added the check, but have not tested the code again. The motivation was to either 
- add an minimal example module for new users there or,
- simply leaving the folder, with an .gitkeep file, for the future

FriedrichAltheide committed:
add nix style to editorconfig

FriedrichAltheide committed:
Only create the module folder if necessary

FriedrichAltheide committed:
Initial test support

FriedrichAltheide committed:
remove old sound.enable setting

FriedrichAltheide committed:
Use substituters from the install env

FriedrichAltheide commented:
@vlinkz @emilazy Do you have further comments / feedback?

FriedrichAltheide committed:
Reapply: Don't emit autoLogin config with eval warning
Fix this evaluation warning on the first rebuild after install:

  [test@nixos:~]$ sudo nixos-rebuild build
  building Nix...
  building the system configuration...
  evaluation warning: The option `services.xserver.displayManager.autoLogin' defined in `/etc/nixos/configuration.nix' has been renamed to `services.displayManager.autoLogin'.

Original commit id: 41f9b1ba0102f378cddee549281fca0579cdc006
Author: Bjørn Forsman <bjorn.forsman@gmail.com>
Co-authored-by: Bjørn Forsman <bjorn.forsman@gmail.com>

ElvishJerricco reviewed (changes requested):
I think we should probably ditch all the `example-custom-iso` stuff. For one thing, it's pointing at your nixpkgs fork and relies on changes that aren't upstream, which is not appropriate. But realistically all of that should be a nixpkgs PR, not bespoke code in this repo.

FriedrichAltheide commented:
@ElvishJerricco Thank you for your feedback.

The `example-custom-iso` was requested out of band by a Nix user who sought guidance on how this pull request could be used and tested.

Currently, I am unable to submit my nixpkgs changes (https://github.com/FriedrichAltheide/nixpkgs/commits/calamares-nixos-extensions-custom-installer/) as a PR from my fork to upstream until this PR is merged. Without my nixpkgs changes within my fork, how should this PR be tested and reviewed?

If removing the example folder will facilitate the merging of this PR, I will do so. However, simply removing the example code because it currently uses my fork does not seem right. The example code cannot use the upstream nixpkgs until this PR is merged, and the required module changes from my nixpkgs fork have been completed and merged. Once this PR and my future module changes PR are merged, the upstream nixpkgs can be used within the `example-custom-iso`. Afterwards, the example within `example-custom-iso` will offer a starting point to users who would like to create their own customised calamares nixOS installer.

FriedrichAltheide committed:
fix tests

FriedrichAltheide committed:
remove trailing spaces

FriedrichAltheide commented:
ping

FriedrichAltheide commented:
ping

FriedrichAltheide commented:
@vlinkz @ElvishJerricco Would one of you please be so kind as to give me a hint as to whether it is worth the time to rebase this PR? Does this PR even have a change to be merged at all?

ElvishJerricco commented:
@FriedrichAltheide Yes, sorry. I would like to see this merged.

As for the example code and the nixpkgs fork: Ideally, this repo ought to have its own reference implementations of the packaging and nixos module that nixos could follow, so that we could develop those things within this repo and port the changes downstream to nixpkgs when the version is bumped. Then we wouldn't need the example code to point at a nixpkgs fork.

But, that's out of scope for this PR. In the meantime, I think I would suggest housing the example in the nixpkgs fork as documentation in the nixos manual, rather than pointing calamares-nixos-extensions at a fork. Then the example becomes part of the docs in the PR adding these features to nixos.

Other than that, I think I'm pretty happy with how this PR looks already.

FriedrichAltheide committed:
move example code to flake.nix

create reference module / package

FriedrichAltheide commented:
@ElvishJerricco Please have a look at the recent changes. I have added a reference packaging and module in `nix/calamares-nixos-extensions`.

The flake.nix was extended by the following packages:
- `calamares-nixos-extensions`: Build the calamares-nixos-extensions using the reference packaging
- `defaultCalamaresIso`: Build an ISO using the reference module and packaging without customisations
- `customCalamaresIso`: Build and ISO using the reference module and packaging with some customisations

The GitHub workflow was extended to build `.#calamares-nixos-extensions`, `.#defaultCalamaresIso` and `.#customCalamaresIso`.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-preseed-configuration/62984/5

FriedrichAltheide commented:
@ElvishJerricco ping

ElvishJerricco commented on modules/nixos/main.py:
```diff
@@ -23,316 +23,44 @@
+    libcalamares.utils.error('Failed to load config snippet {}'.format(snippetName))
+    return ""
+
+def copyModules(root_mount_point, subpath):
+    srcModuleFolderPath = '{}/{}/.'.format(getCalamaresNixosModulePath(), subpath)
+    destModuleFolderPath = '{}/etc/nixos/modules/'.format(root_mount_point)
+    if os.path.isdir(srcModuleFolderPath) and len(os.listdir(srcModuleFolderPath)) != 0:
+        if not os.path.isdir(destModuleFolderPath):
+            libcalamares.utils.host_env_process_output(
+                ["mkdir", "-p", destModuleFolderPath], None
+            )
+        libcalamares.utils.debug('Copying module directory {}'.format(srcModuleFolderPath))
+        libcalamares.utils.host_env_process_output(["cp", "-r", srcModuleFolderPath, destModuleFolderPath], None)
+    else:
+        libcalamares.utils.debug('Skipping non-existing module directory {}'.format(srcModuleFolderPath))
```
I don't know why but this still seems to be creating an empty `/etc/nixos/modules` directory.

ElvishJerricco commented on modules/nixos/main.py:
```diff
@@ -23,316 +23,44 @@
-cfghead = """# Edit this configuration file to define what should be installed on
-# your system.  Help is available in the configuration.nix(5) man page
-# and in the NixOS manual (accessible by running ‘nixos-help’).
+    # Check if a custom snippet exists for the config name
+    snippetConfigPath = '{}/defaultConfigs/snippets/{}.snippet'.format(configPath, snippetName)
+    customConfigFilePath = '{}/customConfigs/snippets/{}.snippet'.format(configPath, snippetName)
```
I'm not a fan of having these two `defaultConfigs` and `customConfigs` things. I like the way systemd does this sort of thing better, where a file in `/usr/lib/foo` is generally overridden by a like file in `/run/foo` which is also overridden by a like file in `/etc/foo`. In Nix the `/usr/lib/` stuff is generally replaced with `/nix/store/...-foo/` for the same purpose.

So I think it would be better if the package just looked in its own output path, but allowed files in a directory like `/etc/calamares-nixos-extensions/` or something like that to override. Then you don't have to do this awkward dance of overriding the package to change stuff like this, and you can just put custom configs in `environment.etc` or something.

ElvishJerricco reviewed (changes requested):
Hey @FriedrichAltheide, I'm so sorry this is taking so long. Frankly, this repo is basically in maintenance mode, and I'm not very well able to make time for larger PRs like this. I have left some more thoughts, which I should have caught my first time around reviewing this; sorry about that.

I hope you keep contributing here; we do need to find someone to take on a more active role in this repo.

ElvishJerricco commented:
This repo has been merged upstream into nixpkgs and will be archived. I'd love to see this work merged in some form, so please feel free to open a PR on nixpkgs to continue it.
